fl4health/utils/client.py

```diff
@@ -0,0 +1,130 @@
+import copy
+from typing import Dict, Iterable, TypeVar
+
+import torch
+import torch.nn as nn
+from flwr.common.logger import LOG_COLORS
+from flwr.common.typing import Any, Scalar
+from tqdm import tqdm
+
+from fl4health.utils.logging import LoggingMode
+from fl4health.utils.metrics import MetricPrefix
+from fl4health.utils.typing import TorchInputType, TorchTargetType
+
+T = TypeVar("T", TorchInputType, TorchTargetType)
+
+
+def fold_loss_dict_into_metrics(
+    metrics: Dict[str, Scalar], loss_dict: Dict[str, float], logging_mode: LoggingMode
+) -> None:
+    # Prefixing the loss value keys with the mode from which they are generated
+    if logging_mode is LoggingMode.VALIDATION:
+        metrics.update({f"{MetricPrefix.VAL_PREFIX.value} {key}": loss_val for key, loss_val in loss_dict.items()})
+    else:
+        metrics.update({f"{MetricPrefix.TEST_PREFIX.value} {key}": loss_val for key, loss_val in loss_dict.items()})
+
+
+def move_data_to_device(data: T, device: torch.device) -> T:
+    """
+    _summary_
+
+    Args:
+        data (T): The data to move to self.device. Can be a TorchInputType or a TorchTargetType
+        device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+            'cuda'
+
+    Raises:
+        TypeError: Raised if data is not one of the types specified by TorchInputType or TorchTargetType
+
+    Returns:
+        T: The data argument except now it's been moved to self.device
+    """
+    # Currently we expect both inputs and targets to be either tensors
+    # or dictionaries of tensors
+    if isinstance(data, torch.Tensor):
+        return data.to(device)
+    elif isinstance(data, dict):
+        return {key: value.to(device) for key, value in data.items()}
+    else:
+        raise TypeError(
+            "data must be of type torch.Tensor or Dict[str, torch.Tensor]. \
+                If definition of TorchInputType or TorchTargetType has \
+                changed this method might need to be updated or split into \
+                two"
+        )
+
+
+def is_empty_batch(input: TorchInputType) -> bool:
```

Comment:
nit: We are checking if the batch is empty but we are also checking if the input is valid (ie for dictionaries, verifying all tensors have same length. Is it worth a slight renaming of the function?

Reply:
Yeah I think that makes sense. Will rename.